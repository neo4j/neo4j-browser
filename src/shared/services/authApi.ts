import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import type { AuthenticationMethod } from 'shared/types/auth'
import type { FetchBaseQueryError } from '@reduxjs/toolkit/query'

interface Credentials {
  host: string
  username?: string
  password?: string
  authenticationMethod: AuthenticationMethod
  encrypted?: boolean
}

interface AuthResponse {
  success: boolean
  error?: string
  token?: string
  authUrl?: string
  requiresPasswordChange?: boolean
  SSOProviders?: Array<{
    id: string
    name: string
    authUrl: string
  }>
}

export const authApi = createApi({
  reducerPath: 'authApi',
  baseQuery: fetchBaseQuery({ 
    baseUrl: '/api',
    credentials: 'include',
  }),
  endpoints: (builder) => ({
    connect: builder.mutation<AuthResponse, Credentials>({
      query: (credentials) => ({
        url: '/connect',
        method: 'POST',
        body: credentials,
        headers: {
          'Content-Type': 'application/json',
          'X-CSRF-Token': document.cookie.match(/csrf-token=([^;]+)/)?.[1] || ''
        }
      })
    }),
    verifyCredentials: builder.mutation<AuthResponse, Credentials>({
      query: (credentials) => ({
        url: '/verify-credentials',
        method: 'POST',
        body: credentials
      })
    }),
    getSSOProviders: builder.query<AuthResponse, string>({
      query: (host) => `/sso-providers?host=${encodeURIComponent(host)}`
    }),
    logout: builder.mutation<void, void>({
      query: () => ({
        url: '/logout',
        method: 'POST'
      })
    }),
    ssoLogin: builder.mutation<AuthResponse, void>({
      query: () => ({
        url: '/sso/login',
        method: 'POST'
      }),
      async onQueryStarted(_, { queryFulfilled }) {
        try {
          const { data } = await queryFulfilled
          if (data.success) {
            window.location.href = data.authUrl || '/'
          }
        } catch (err) {
          console.error('SSO login failed:', err)
        }
      }
    })
  })
})

export const {
  useConnectMutation,
  useVerifyCredentialsMutation,
  useGetSSOProvidersQuery,
  useLogoutMutation,
  useSsoLoginMutation
} = authApi 