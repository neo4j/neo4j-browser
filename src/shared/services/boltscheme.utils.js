/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

const BOLT_DIRECT_SCHEME = 'bolt'
const BOLT_ROUTING_SCHEME = 'neo4j'

export const isNonSupportedRoutingSchemeError = e =>
  e.code === 'ServiceUnavailable' &&
  e.message.includes('Could not perform discovery')

export const isNonRoutingScheme = (url = '') =>
  typeof url === 'string' && url.startsWith(`${BOLT_DIRECT_SCHEME}://`)

export const toNonRoutingScheme = url =>
  typeof url === 'string' &&
  `${BOLT_DIRECT_SCHEME}${getSchemeFlag(url)}://${stripScheme(url)}`

export const getScheme = url => {
  if (!url) {
    return ''
  }
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = url.split('://')
  return scheme
}

export const stripScheme = url => {
  const [_scheme, ...rest] = (url || '').split('://')
  if (!rest || !rest.length) {
    return _scheme
  }
  return rest.join('://')
}

export const isSecureBoltScheme = url => {
  if (url && !url.includes('://')) {
    return false
  }
  const [scheme] = (url || '').split('://')
  if (!scheme) {
    return false
  }
  return scheme.endsWith('+s') || scheme.endsWith('+ssc')
}

export const getSchemeFlag = (url = '') => {
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = (url || '').split('://')
  if (!scheme.includes('+')) {
    return ''
  }
  return `+${scheme.split('+').pop()}`
}
const stripSchemeFlag = url => {
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = (url || '').split('://')
  if (!scheme.includes('+')) {
    return scheme
  }
  return scheme.split('+')[0]
}

const toggleSchemeSecurity = url => {
  if (url && !url.includes('://')) {
    return url
  }
  if (isSecureBoltScheme(url)) {
    return stripSchemeFlag(url)
  }
  return `${getScheme(url)}+s`
}

export const toggleSchemeRouting = (url = '') => {
  if (!url) {
    return ''
  }
  if (url && !url.includes('://')) {
    return url
  }

  if (url.startsWith(BOLT_ROUTING_SCHEME)) {
    return url.replace(BOLT_ROUTING_SCHEME, BOLT_DIRECT_SCHEME)
  }
  if (url.startsWith(BOLT_DIRECT_SCHEME)) {
    return url.replace(BOLT_DIRECT_SCHEME, BOLT_ROUTING_SCHEME)
  }
  return url
}

export const generateBoltUrl = (allowedSchemes, url, fallbackScheme) => {
  const rewrites = {
    'bolt+routing://': `${BOLT_ROUTING_SCHEME}://`
  }

  if (!url || typeof url !== 'string') {
    if (
      allowedSchemes &&
      fallbackScheme &&
      allowedSchemes.includes(fallbackScheme)
    ) {
      return `${fallbackScheme}://`
    }
    let scheme = allowedSchemes ? allowedSchemes[0] : fallbackScheme || 'neo4j'
    return `${scheme}://`
  }

  // Rewrite any alias. Break on first hit.
  for (const candidate in rewrites) {
    if (url.startsWith(candidate)) {
      url = `${rewrites[candidate]}${stripScheme(url)}`
      break
    }
  }

  const scheme = getScheme(url)

  // We accept all schemes
  if (!allowedSchemes) {
    if (scheme) {
      return url
    }
    return `${BOLT_ROUTING_SCHEME}://${url}`
  }

  if (!scheme && fallbackScheme && allowedSchemes.includes(fallbackScheme)) {
    return `${fallbackScheme}://${url}`
  }

  // Input scheme allowed
  if (allowedSchemes.includes(scheme)) {
    return url
  }

  // Scheme not allowed, can we toggle encryption to allow it?
  if (scheme) {
    const toggledSecurityScheme = toggleSchemeSecurity(url)
    if (allowedSchemes.includes(toggledSecurityScheme)) {
      return `${toggledSecurityScheme}://${stripScheme(url)}`
    }
  }

  // Either no scheme entered or we can't guess an allowed, let's override it with something default
  const defaultScheme =
    fallbackScheme && allowedSchemes.includes(fallbackScheme)
      ? fallbackScheme
      : allowedSchemes[0]
  return `${defaultScheme}://${stripScheme(url)}`
}
