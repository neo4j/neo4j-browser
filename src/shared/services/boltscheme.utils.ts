/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { INSECURE_SCHEMES, SECURE_SCHEMES } from 'shared/modules/app/appDuck'

const BOLT_DIRECT_SCHEME = 'bolt'
const BOLT_ROUTING_SCHEME = 'neo4j'

export const isNonSupportedRoutingSchemeError = (e: {
  code: string
  message: string
}) =>
  e.code === 'ServiceUnavailable' &&
  (e.message.includes('Could not perform discovery') ||
    e.message.includes('routing'))

export const isNonRoutingScheme = (url = '') =>
  typeof url === 'string' && url.startsWith(`${BOLT_DIRECT_SCHEME}://`)

export const toNonRoutingScheme = (url: string) =>
  typeof url === 'string' &&
  `${BOLT_DIRECT_SCHEME}${getSchemeFlag(url)}://${stripQueryString(
    stripScheme(url)
  )}`

export const getScheme = (url: string) => {
  if (!url) {
    return ''
  }
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = url.split('://')
  return scheme
}

export const stripScheme = (url: string) => {
  const [_scheme, ...rest] = (url || '').split('://')
  if (!rest || !rest.length) {
    return _scheme
  }
  return rest.join('://')
}

export const boltUrlsHaveSameHost = (host1: string, host2: string): boolean =>
  stripScheme(stripQueryString(host1)) === stripScheme(stripQueryString(host2))

export const stripQueryString = (url: string) => {
  const [_host] = (url || '').split('?')
  return _host
}

export const isSecureBoltScheme = (url: string) => {
  if (url && !url.includes('://')) {
    return false
  }
  const [scheme] = (url || '').split('://')
  if (!scheme) {
    return false
  }
  return scheme.endsWith('+s') || scheme.endsWith('+ssc')
}

export const getSchemeFlag = (url = '') => {
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = (url || '').split('://')
  if (!scheme.includes('+')) {
    return ''
  }
  return `+${scheme.split('+').pop()}`
}
const stripSchemeFlag = (url: string) => {
  if (url && !url.includes('://')) {
    return ''
  }
  const [scheme] = (url || '').split('://')
  if (!scheme.includes('+')) {
    return scheme
  }
  return scheme.split('+')[0]
}

const toggleSchemeSecurity = (url: string) => {
  if (url && !url.includes('://')) {
    return url
  }
  if (isSecureBoltScheme(url)) {
    return stripSchemeFlag(url)
  }
  return `${getScheme(url)}+s`
}

export const toggleSchemeRouting = (url = '') => {
  if (!url) {
    return ''
  }
  if (url && !url.includes('://')) {
    return url
  }

  if (url.startsWith(BOLT_ROUTING_SCHEME)) {
    return url.replace(BOLT_ROUTING_SCHEME, BOLT_DIRECT_SCHEME)
  }
  if (url.startsWith(BOLT_DIRECT_SCHEME)) {
    return url.replace(BOLT_DIRECT_SCHEME, BOLT_ROUTING_SCHEME)
  }
  return url
}

export const boltToHttp = (boltHost: string): string => {
  const withOutScheme = stripScheme(boltHost)
  if (SECURE_SCHEMES.some(scheme => boltHost.startsWith(scheme))) {
    return `https://${withOutScheme}`
  }
  if (INSECURE_SCHEMES.some(scheme => boltHost.startsWith(scheme))) {
    return `http://${withOutScheme}`
  }
  return boltHost
}

export const generateBoltUrl = (
  allowedSchemes: string[],
  url: string,
  fallbackScheme?: string
) => {
  const rewrites: Record<string, string> = {
    'bolt+routing://': `${BOLT_ROUTING_SCHEME}://`
  }

  if (!url || typeof url !== 'string') {
    if (
      allowedSchemes &&
      fallbackScheme &&
      allowedSchemes.includes(fallbackScheme)
    ) {
      return `${fallbackScheme}://`
    }
    const scheme = allowedSchemes
      ? allowedSchemes[0]
      : fallbackScheme || 'neo4j'
    return `${scheme}://`
  }

  // Rewrite any alias. Break on first hit.
  for (const candidate in rewrites) {
    if (url.startsWith(candidate)) {
      url = `${rewrites[candidate]}${stripScheme(url)}`
      break
    }
  }

  const scheme = getScheme(url)

  // We accept all schemes
  if (!allowedSchemes) {
    if (scheme) {
      return url
    }
    return `${BOLT_ROUTING_SCHEME}://${url}`
  }

  if (!scheme && fallbackScheme && allowedSchemes.includes(fallbackScheme)) {
    return `${fallbackScheme}://${url}`
  }

  // Input scheme allowed
  if (allowedSchemes.includes(scheme)) {
    return url
  }

  // Scheme not allowed, can we toggle encryption to allow it?
  if (scheme) {
    const toggledSecurityScheme = toggleSchemeSecurity(url)
    if (allowedSchemes.includes(toggledSecurityScheme)) {
      return `${toggledSecurityScheme}://${stripScheme(url)}`
    }
  }

  // Either no scheme entered or we can't guess an allowed, let's override it with something default
  const defaultScheme =
    fallbackScheme && allowedSchemes.includes(fallbackScheme)
      ? fallbackScheme
      : allowedSchemes[0]
  return `${defaultScheme}://${stripScheme(url)}`
}
