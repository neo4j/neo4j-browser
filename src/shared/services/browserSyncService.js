/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/database'

const appIsInitialized = () => firebase.apps.length && firebase.apps.length > 0

export const authenticate = (dataToken, onSignout = null) => {
  return firebase
    .auth()
    .signInWithCustomToken(dataToken)
    .then(a => {
      firebase.auth().onAuthStateChanged(user => {
        if (!user || !user.uid) {
          onSignout && onSignout()
        }
      })
      return a
    })
}

export const initialize = config => {
  if (appIsInitialized()) {
    return
  }

  return firebase.initializeApp(config)
}

export const status = () => {
  return firebase.database().ref('.info/connected')
}

export const getResourceFor = userId => {
  return firebase.database().ref(`users/${userId}`)
}

export const syncResourceFor = (userId, key, value) => {
  const userRef = firebase.database().ref(`users/${userId}`)
  userRef.child(key).set(value)
}

export const setupUser = (userId, initialData) => {
  firebase
    .database()
    .ref(`users/${userId}`)
    .set(initialData)
}

export const signOut = () => {
  if (appIsInitialized()) {
    firebase.auth().signOut()
  }
}
