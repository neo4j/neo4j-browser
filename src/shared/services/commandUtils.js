/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { includes, last, split, startsWith } from 'lodash-es'
import { extractStatements } from 'cypher-codemirror'

export function cleanCommand(cmd) {
  const noComments = stripCommandComments(cmd)
  const noEmptyLines = stripEmptyCommandLines(noComments)
  return noEmptyLines
}

export function stripEmptyCommandLines(str) {
  const skipEmptyLines = e => !/^\s*$/.test(e)
  return str
    .split('\n')
    .filter(skipEmptyLines)
    .join('\n')
}

export function stripCommandComments(str) {
  return str
    .split('\n')
    .filter(line => !line.trim().startsWith('//'))
    .join('\n')
}

export function splitStringOnFirst(str, delimiter) {
  const parts = str.split(delimiter)
  return [].concat(parts[0], parts.slice(1).join(delimiter))
}

export function splitStringOnLast(str, delimiter) {
  const parts = str.split(delimiter)
  return [].concat(
    parts.slice(0, parts.length - 1).join(delimiter),
    parts[parts.length - 1]
  )
}

export const isCypherCommand = cmd => {
  const cleanCmd = cleanCommand(cmd)
  return cleanCmd[0] !== ':'
}

export const buildCommandObject = (action, interpret) => {
  const interpreted = getInterpreter(interpret, action.cmd, action.ignore)
  return { action, interpreted, useDb: action.useDb }
}

export const getInterpreter = (interpret, cmd, ignore = false) => {
  if (ignore) return interpret('noop')
  if (isCypherCommand(cmd)) return interpret('cypher')
  return interpret(cleanCommand(cmd).substr(1))
}

export const isNamedInterpreter = interpreter =>
  interpreter && interpreter.name !== 'catch-all'

export const extractPostConnectCommandsFromServerConfig = str => {
  const substituteStr = '@@semicolon@@'
  const substituteRe = new RegExp(substituteStr, 'g')
  const replaceFn = m => m.replace(/;/g, substituteStr)
  const qs = [/(`[^`]*?`)/g, /("[^"]*?")/g, /('[^']*?')/g]
  qs.forEach(q => (str = str.replace(q, replaceFn)))
  const splitted = str
    .split(';')
    .map(item => item.trim())
    .map(item => item.replace(substituteRe, ';'))
    .filter(item => item && item.length)
  return splitted && splitted.length ? splitted : undefined
}

const getHelpTopic = str => splitStringOnFirst(str, ' ')[1] || 'help' // Map empty input to :help help
const stripPound = str => splitStringOnFirst(str, '#')[0]
const lowerCase = str => str.toLowerCase()
const trim = str => str.trim()
const replaceSpaceWithDash = str => str.replace(/\s/g, '-')
const snakeToCamel = str =>
  str.replace(/(-\w)/g, match => match[1].toUpperCase())
const camelToSnake = (name, separator) => {
  return name
    .replace(
      /([a-z]|(?:[A-Z0-9]+))([A-Z0-9]|$)/g,
      (_, $1, $2) => $1 + ($2 && (separator || '_') + $2)
    )
    .toLowerCase()
}

export const transformCommandToHelpTopic = inputStr =>
  [inputStr || '']
    .map(stripPound)
    .map(getHelpTopic)
    .map(lowerCase)
    .map(trim)
    .map(replaceSpaceWithDash)
    .map(snakeToCamel)[0]

export const transformHelpTopicToCommand = inputStr => {
  if (inputStr.indexOf('-') > -1) {
    return inputStr
  }
  return camelToSnake(inputStr, '-').replace('-', ' ')
}

const quotedRegex = /^"(.*)"|'(.*)'/
const arrowFunctionRegex = /^.*=>\s*([^$]*)$/

export const mapParamToCypherStatement = (key, param) => {
  const quotedKey = key.match(quotedRegex)
  const cleanKey = quotedKey
    ? `\`${quotedKey[1]}\``
    : typeof key !== 'string'
    ? `\`${key}\``
    : key
  const returnAs = value => `RETURN ${value} as ${cleanKey}`

  const matchParamFunction = param.toString().match(arrowFunctionRegex)
  if (matchParamFunction) {
    return returnAs(matchParamFunction[1])
  }
  return returnAs(param)
}

export const extractStatementsFromString = str => {
  const cleanCmd = cleanCommand(str)
  const parsed = extractStatements(cleanCmd)
  const { statements } = parsed.referencesListener
  return statements[0]
    .raw()
    .map(stmt => stmt.getText().trim())
    .filter(_ => _)
}

export const getCommandAndParam = str => {
  const [serverCmd, props] = splitStringOnFirst(
    splitStringOnFirst(str, ' ')[1],
    ' '
  )
  return [serverCmd, props]
}

export function tryGetRemoteInitialSlideFromUrl(url) {
  const hashBang = includes(url, '#') ? last(split(url, '#')) : ''

  if (!startsWith(hashBang, 'slide-')) return 0

  const slideIndex = Number(last(split(hashBang, 'slide-')))

  return !isNaN(slideIndex) ? slideIndex : 0
}
