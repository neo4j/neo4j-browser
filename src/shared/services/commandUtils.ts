/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { extractStatements } from '@neo4j-cypher/extract-statements'

export function cleanCommand(cmd: string): string {
  const noComments = stripCommandComments(cmd)
  return stripEmptyCommandLines(noComments)
}

export function stripEmptyCommandLines(str: string): string {
  const skipEmptyLines = (e: string) => !/^\s*$/.test(e)
  return str.split('\n').filter(skipEmptyLines).join('\n')
}

export function stripCommandComments(str: string): string {
  return str
    .split('\n')
    .filter((line: string) => !line.trim().startsWith('//'))
    .join('\n')
}

export function splitStringOnFirst(
  str: string,
  delimiter: string | RegExp
): string[] {
  const parts = str.split(delimiter)
  const strWithoutFirst = str.slice(parts[0].length + 1)
  return ([] as string[]).concat(parts[0], strWithoutFirst)
}

export function splitStringOnLast(
  str: string,
  delimiter: string | RegExp
): string[] {
  const parts = str.split(delimiter)
  const lastPart = parts[parts.length - 1]
  const stringWithoutLast =
    parts.length > 1 ? str.slice(0, str.length - lastPart.length - 1) : ''
  return ([] as string[]).concat(stringWithoutLast, lastPart)
}

export const isCypherCommand = (cmd: string): boolean => {
  const cleanCmd = cleanCommand(cmd)
  return cleanCmd[0] !== ':'
}

export const buildCommandObject = (action: any, interpret: any) => {
  const interpreted = getInterpreter(interpret, action.cmd, action.ignore)
  return { action, interpreted, useDb: action.useDb }
}

export const getInterpreter = (interpret: any, cmd: string, ignore = false) => {
  if (ignore) return interpret('noop')
  if (isCypherCommand(cmd)) return interpret('cypher')
  return interpret(cleanCommand(cmd).substr(1))
}

export const extractPostConnectCommandsFromServerConfig = (
  str: string
): string[] | undefined => {
  const substituteStr = '@@semicolon@@'
  const substituteRe = new RegExp(substituteStr, 'g')
  const replaceFn = (m: string) => m.replace(/;/g, substituteStr)
  const qs = [/(`[^`]*?`)/g, /("[^"]*?")/g, /('[^']*?')/g]
  qs.forEach(q => (str = str.replace(q, replaceFn)))
  const splitted = str
    .split(';')
    .map((item: string) => item.trim())
    .map((item: string) => item.replace(substituteRe, ';'))
    .filter((item: string) => item && item.length)
  return splitted && splitted.length ? splitted : undefined
}

const getHelpTopic = (str: string) => splitStringOnFirst(str, ' ')[1] || 'help' // Map empty input to :help help
const stripPound = (str: string) => splitStringOnFirst(str, '#')[0]
const lowerCase = (str: string) => str.toLowerCase()
const trim = (str: string) => str.trim()
const replaceSpaceWithDash = (str: string) => str.replace(/\s/g, '-')
const snakeToCamel = (str: string) =>
  str.replace(/(-\w)/g, (match: string) => match[1].toUpperCase())
const camelToSnake = (name: string, separator: string) => {
  return name
    .replace(
      /([a-z]|(?:[A-Z0-9]+))([A-Z0-9]|$)/g,
      (_: string, $1: string, $2: string) =>
        $1 + ($2 && (separator || '_') + $2)
    )
    .toLowerCase()
}

export const transformCommandToHelpTopic = (inputStr?: string): string =>
  [inputStr || '']
    .map(stripPound)
    .map(getHelpTopic)
    .map(lowerCase)
    .map(trim)
    .map(replaceSpaceWithDash)
    .map(snakeToCamel)[0]

export const transformHelpTopicToCommand = (inputStr: string): string => {
  if (inputStr.indexOf('-') > -1) {
    return inputStr
  }
  return camelToSnake(inputStr, '-').replace('-', ' ')
}

const quotedRegex = /^"(.*)"|'(.*)'/
const arrowFunctionRegex = /^.*=>\s*([^$]*)$/

export const mapParamToCypherStatement = (key: any, param: any): string => {
  const quotedKey = key.match(quotedRegex)
  const cleanKey = quotedKey
    ? `\`${quotedKey[1]}\``
    : typeof key !== 'string'
    ? `\`${key}\``
    : key
  const returnAs = (value: any) => `RETURN ${value} as ${cleanKey}`

  const matchParamFunction = param.toString().match(arrowFunctionRegex)
  if (matchParamFunction) {
    return returnAs(matchParamFunction[1])
  }
  return returnAs(param)
}

export const extractStatementsFromString = (str: string) => {
  const cleanCmd = cleanCommand(str)
  const parsed = extractStatements(cleanCmd)
  const { statements } = parsed.referencesListener
  return statements[0]
    .raw()
    .map((stmt: any) => stmt.getText().trim())
    .filter((_: any) => _)
}

export const getCommandAndParam = (str: string): string[] => {
  const [serverCmd, props] = splitStringOnFirst(
    splitStringOnFirst(str, ' ')[1],
    ' '
  )
  return [serverCmd, props]
}
