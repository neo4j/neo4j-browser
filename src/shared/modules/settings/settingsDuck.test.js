/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import reducer, {
  DISABLE_IMPLICIT_INIT_COMMANDS,
  NAME,
  UPDATE,
  REPLACE,
  shouldReportUdc,
  getInitCmd
} from './settingsDuck'
import { dehydrate } from 'services/duckUtils'

describe('settings reducer', () => {
  test('handles initial value', () => {
    const nextState = dehydrate(reducer(undefined, { type: '' }))
    expect(nextState.cmdchar).toEqual(':')
  })

  test('handles UPDATE without initial state', () => {
    const action = {
      type: UPDATE,
      state: {
        greeting: 'hello'
      }
    }
    const nextState = reducer(undefined, action)
    expect(nextState.greeting).toEqual('hello')
  })

  test('handles UPDATE', () => {
    const initialState = { cmdchar: ':', greeting: 'hello', type: 'human' }
    const action = {
      type: UPDATE,
      state: {
        greeting: 'woff',
        type: 'dog'
      }
    }
    const nextState = dehydrate(reducer(initialState, action))
    expect(nextState.cmdchar).toEqual(':')
    expect(nextState.greeting).toEqual('woff')
    expect(nextState.type).toEqual('dog')
  })
  test('handles REPLACE', () => {
    const initialState = { greeting: 'hello', type: 'human' }
    const action = {
      type: REPLACE,
      state: {
        new: 'conf'
      }
    }
    const nextState = dehydrate(reducer(initialState, action))
    expect(nextState.greeting).toBeUndefined()
    expect(nextState.type).toBeUndefined()
    expect(nextState).toMatchSnapshot()
  })

  it('defaults playImplicitInitCommands to true', () => {
    expect(reducer(undefined, { type: 'dummy action' })).toEqual(
      expect.objectContaining({ playImplicitInitCommands: true })
    )
  })

  it('sets playImplicitInitCommands to false on DISABLE_IMPLICIT_INIT_COMMANDS', () => {
    expect(
      reducer(undefined, { type: DISABLE_IMPLICIT_INIT_COMMANDS })
    ).toEqual(expect.objectContaining({ playImplicitInitCommands: false }))
  })
})

describe('Selectors', () => {
  test('shouldReportUdc casts to true for anything not false', () => {
    // Given
    const tests = [
      { test: true, expect: true },
      { test: 1, expect: true },
      { test: '1', expect: true },
      { test: 'on', expect: true },
      { test: null, expect: true },
      { test: undefined, expect: true },
      { test: false, expect: false }
    ]

    // When && Then
    tests.forEach(t => {
      const state = {
        [NAME]: { shouldReportUdc: t.test }
      }
      expect(shouldReportUdc(state)).toEqual(t.expect)
    })
  })
  test("let getInitCmd be falsy and cast to empty string if that's the case", () => {
    // Given
    const tests = [
      { test: ':play start', expect: ':play start' },
      { test: null, expect: '' },
      { test: undefined, expect: '' },
      { test: '', expect: '' },
      { test: ' ', expect: '' },
      {
        test: '//Todays number is:\nRETURN rand()',
        expect: '//Todays number is:\nRETURN rand()'
      }
    ]

    // When && Then
    tests.forEach(t => {
      const state = {
        [NAME]: { initCmd: t.test }
      }
      expect(getInitCmd(state)).toEqual(t.expect)
    })
  })
})
