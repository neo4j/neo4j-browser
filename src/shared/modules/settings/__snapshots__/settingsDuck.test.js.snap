// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings reducer handles REPLACE 1`] = `
Object {
  "autoComplete": true,
  "browserSyncDebugServer": null,
  "cmdchar": ":",
  "codeFontLigatures": true,
  "connectionTimeout": 30000,
  "editorAutocomplete": true,
  "editorLint": false,
  "enableMultiStatementMode": true,
  "initCmd": ":play start",
  "initialNodeDisplay": 300,
  "maxFieldItems": 500,
  "maxFrames": 30,
  "maxHistory": 30,
  "maxNeighbours": 100,
  "maxRows": 1000,
  "new": "conf",
  "playImplicitInitCommands": true,
  "scrollToTop": true,
  "shouldReportUdc": true,
  "showSampleScripts": true,
  "theme": "auto",
  "useCypherThread": true,
}
`;
